pilot/ARCHITECTURE.md

```diff
@@ -0,0 +1,181 @@
+        p(Pods)
+        ksi(ServiceInstances)
+        kwi(WorkloadInstances)
+        s-->ksi
+        e-->ksi
+        p-->kwi
+    end
+    subgraph ServiceEntry Controller
+        se(ServiceEntry)
+        we(WorkloadEntry)
+        ssi(ServiceInstances)
+        swi(WorkloadInstances)
+        se-->ssi
+        swi-->ssi
+        we-->swi
+    end
+    kwi-->ssi
+    swi-->ksi
+```
+
+For the most part this is fairly straight forward. However, we support `ServiceEntry` selecting `Pod`, and `Service` selecting `WorkloadEntry`, which leads to cross-controller communication.
+
+Note: the asymmetry with `Pods` not contributing to Kube controller's `ServiceInstances` is due to the use of `Endpoints` which itself is derived from `Pod` from Kubernetes core.
+
+#### PushContext
+
+`PushContext` is an immutable snapshot of the current state of the world. It is regenerated (usually partially) on each configuration push (more on this below). Due to being a snapshot, most lookups are lock-free.
+
+`PushContext` is built up by querying the above layers. For some simple use cases, this is as simple as storing something like `configstore.List(SomeType)`; in this case, the only difference from directly exposing the configstore is to snapshot the current state. In other cases, some pre-computations and indexes are computed to make later accesses efficient.
+
+#### Endpoints
+
+Endpoints have an optimized code path, as they are by far the most frequently updated resource - in a steady cluster, this will often be the *only* change, caused by scale up/down.
+
+As a result, they do not go through `PushContext`, and changes do not trigger a `PushContext` recomputation. Instead, the current state is incrementally computed based on events from `ServiceDiscovery`.
+
+#### Conclusion
+
+Overall, the high level config ingestion flow:
+
+```mermaid
+graph TD
+    sd(Service Discovery)
+    cs(ConfigStore)
+    ep(Endpoints)
+    pc(PushContext)
+    sd-->pc
+    cs-->pc
+    sd-->ep
+```
+
+### Config Translation
+
+Config Translation turns the above inputs into the actual types consumed by the connected XDS clients (typically Envoy). This is done by `Generators`, which register a function to build a given type. For example, there is a `RouteGenerator` responsible for building `Routes`. Along with the core Envoy XDS types, there are a few custom Istio types, such as our `NameTable` type used for DNS, as well as debug interfaces.
+
+`Generators` get as input the `Proxy` (a representation of the current client), the current `PushContext` snapshot, and a list of config updates that caused the change.
+
+The `Proxy` as an input parameter is important, and a major distinction from some other XDS implementations. We are not able to statically translate inputs to XDS without per-client information. For example, we rely on the client's labels to determine the set of policies applied. While this is necessary to implement Istio's APIs, it does limit performance substantially.
```

Comment:
I would add 'new k8s Gateway and ambient API surfaces are less dependent on per-client information' - and maybe make it more clear that this model is at least controversial and creates a lot of complexity and scale limits that we may be able to avoid as API usage shifts.